src/util/mogps.jl

```diff
@@ -0,0 +1,112 @@
+"""
+    MOutput(x::AbstractVector, out_dim::Integer)
+
+A data type to handle multi-dimensional outputs.
+"""
+struct MOutput{T<:AbstractVector} <: AbstractVector{eltype(eltype(T))} where T <: Real
```

Comment:
I would suggest
```suggestion
struct MOutput{T<:Real,X<:AbstractVector} <: AbstractVector{T}
```
here - does the current definition actually work? Also, I'm not sure if it is useful to wrap a matrix in this struct as well - e.g., the computation of the indices (in the flattened form) seems unnecessarily complicated for matrices since there you could just use the standard linear indexing anyways.

Reply:
Haha yeah I am also surprised it just work.
In your suggested change, `T` is not inferred from the data...

Reply:
That would be part of the constructors, not of the type definition.

Reply:
If it would only wrap vectors of vectors, one could of course do stuff like
```suggestion
struct MOutput{T<:Real,X<:AbstractVector{<:AbstractVector{T}}} <: AbstractVector{T}
```
but the suggestion above would be more generally applicable, so I'm not sure if it's useful to be that specific here.

Reply:
BTW I'm not sure we should define wrappers for vectors of vectors from scratch here anyway if there are existing solutions such as https://github.com/SciML/RecursiveArrayTools.jl#vectorofarray.

Reply:
But here that's not really what we want right? They just make a vector of vector looks like a matrix

Reply:
Oh yeah, I misremembered the VecOfArray thing, I somehow thought it's flattened as well.

Reply:
I think https://github.com/jonniedie/ComponentArrays.jl could be a good candidate no? Some of it could actually be useful to recognize what output we are dealing with (named outputs would be amazing :D)